src/object/srv_obj.c

```diff
@@ -41,6 +41,273 @@
+	struct daos_tx_entry	*dte = NULL;
+	int			 rc;
+
+	cont->sc_tx_committing = 1;
+	rc = vos_dtx_list_committable(cont->sc_hdl, &dte);
+	if (rc > 0) {
+		D_ASSERT(dte != NULL);
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+				dte, rc, cont_hdl->sch_pool->spc_map_version);
+		D_FREE(dte);
+	}
+	cont->sc_tx_committing = 0;
+
+	if (!dta->dta_share) {
+		if (rc > 0) {
+			/* Only trigger aggregation in separated ULT. */
+			while (1) {
+				rc = vos_dtx_aggregate(cont->sc_hdl);
+				if (rc != DTX_ACT_AGGREGATE)
+					break;
+
+				ABT_thread_yield();
+			}
+		}
+
+		if (!cont_hdl->sch_cont)
+			ds_cont_put(cont);
+		ds_cont_hdl_put(cont_hdl);
+	}
+	D_FREE(dta);
+}
+
+static void
+ds_tx_aggregate(void *arg)
+{
+	struct ds_tx_args	*dta = (struct ds_tx_args *)arg;
+
+	D_ASSERT(!dta->dta_share);
+
+	while (1) {
+		int	rc;
+
+		rc = vos_dtx_aggregate(dta->dta_cont->sc_hdl);
+		if (rc != DTX_ACT_AGGREGATE)
+			break;
+
+		ABT_thread_yield();
+	}
+
+	if (!dta->dta_cont_hdl->sch_cont)
+		ds_cont_put(dta->dta_cont);
+	ds_cont_hdl_put(dta->dta_cont_hdl);
+	D_FREE(dta);
+}
+
+static void
+ds_tx_abort(uuid_t po_uuid, uuid_t co_uuid, struct daos_tx_id *dti,
+	    daos_unit_oid_t *oid, uint32_t version)
+{
+	struct daos_tx_entry	 dte;
+
+	dte.dte_xid = *dti;
+	dte.dte_oid = *oid;
+	dtx_abort(po_uuid, co_uuid, &dte, 1, version);
+}
+
+static int
+ds_tx_post_handle(struct ds_cont_hdl *cont_hdl, struct ds_cont *cont,
+		  struct daos_tx_handle *dth, struct daos_tx_id *dti,
+		  daos_unit_oid_t *oid, struct daos_tx_id *dti_cos,
+		  int dti_cos_count, int rc, bool rpc_dispatched,
+		  bool inline_dti_cos, bool leader)
+{
+	int	err;
+
+	if (dti_cos != NULL) {
+		if (rpc_dispatched && rc < 0) {
+			/* XXX: Even if we fail to make related modification
+			 *	for some reason, we still need to commit the
+			 *	DTXs for CoS. Because other replica may have
+			 *	already committed them.
+			 */
+			err = vos_dtx_commit(cont->sc_hdl, dti_cos,
+					     dti_cos_count);
+			D_ERROR(DF_UOID" can not commit DTX "DF_DTI
+				" for CoS: rc = %d\n",
+				DP_UOID(*oid), DP_DTI(dti), err);
+		}
+
+		if (!inline_dti_cos)
+			D_FREE(dti_cos);
+	}
+
+	/* If the DTX is started befoe DTX resync operation (for rebuild),
+	 * then it is possbile that the DTX resync ULT may have aborted
+	 * current DTX before remote replica(s) modification by race. So
+	 * let's check DTX status locally before marking as 'committable'.
+	 */
+	if (leader && rc >= 0 && dti->dti_sec <= cont->sc_dtx_resync_time) {
+		err = vos_dtx_check_committable(cont->sc_hdl, NULL, dti, 0,
+						false);
+		if (err < 0)
+			rc = (err == -DER_NONEXIST ? -DER_INPROGRESS : err);
+	}
+
+	err = vos_dtx_end(dth, rc, leader);
+	switch (err) {
+	case DTX_ACT_AGGREGATE: {
+		struct ds_tx_args *dta;
+
+		dta = ds_tx_prepare_args(cont_hdl, cont, false);
+		if (dta == NULL)
+			D_WARN("Not enough RAM for DTX aggregation.\n");
+		else
+			dss_ult_create(ds_tx_aggregate, dta, DSS_ULT_SELF,
```

Comment:
DTX table is per-container, I don't think it'll be helpful to create multiple ULTs to aggregate same container.

Reply:
You are right. I will add flag to prevent multiple ds_tx_aggregate.

Reply:
I am not sure whether worth for a dedicated ULT. It is expected that the DTX aggregation will be triggered relative rare. If we use dedicated ULT, then most of time, such ULT will be idle. 

From this point view, the current implementation is some kinds of optimization for that: we trigger the DTX aggregation ULT only necessary, and such ULT will not stop until all the qualified DTXs have been handled.

Reply:
will we have dedicate aggregate ULT for VOS? if we do, I wonder if you can borrow that.  Btw: why do you think idle ULT will have a lot overhead to concern?  Just a few cpu cycle as I understand.

Reply:
I am not sure about the VOS aggregation. As my understand, we only trigger the aggregation as requirement, need @NiuYawei for confirmation.

On the other hand, DTX aggregation has no direct relationship with VOS aggregation.

As for dedicated aggregation ULT, I do not know how much overhead will be, if it is always busy polling, then it is unnecessary. On the other hand, from logic view, even if we have dedicated ULT, we also need to check the DTX aggregation conditions, then wakeup such ULT when necessary. So the dedicated ULT does not bring too much benefit, right?

Reply:
For VOS aggregation, we create the per-container ULT when aggregation is triggered (single ULT for each container), the ULT will exit after the aggregation done.

Reply:
That is the same as current patch does for DTX aggregation.